---
title: "Callahan PTB Contamination Analysis"
author: "BJC"
date: "11/16/2017"
output: html_document
---

Replication and refinement of a vaginal microbial signature of preterm birth in two racially distinct cohorts of US women, Callahan et al., 2017, https://doi.org/10.1073/pnas.1705899114.

Load needed packages:
```{r load}
library(decontam);packageVersion("decontam")
library(ggplot2);packageVersion("ggplot2")
library(phyloseq);packageVersion("phyloseq")
library(dada2);packageVersion("dada2")
library(phyloseq);packageVersion("phyloseq")
library(reshape2);packageVersion("reshape2")
library(gridExtra);packageVersion("gridExtra")
```

## Identify Contaminants

The ASV tables used here were generated by DADA2 from the raw data as part of the original publication. Appropriate R objects containing the ASV tables and taxonomies generated as part of that study are included alongside this workflow script.

Read in the DADA2-processed ASV tables and taxonomnic assignments:
```{r}
path <- "~/Desktop/Contamination/Analyses/CallahanPTB" # CHANGE ME
load(file.path(path, "preg_w_negs.rda"))
taxn <- readRDS(file.path(path, "taxn.rds"))
dfn$Negative <- dfn$BodySite %in% c("EC", "Mock", "NTC")
tab <- table(dfn$Run, dfn$Negative); tab # Drop Run04 as it has no vaginal samples
dfn <- dfn[dfn$Run %in% rownames(tab)[rowSums(tab>0)==2],]
stn <- stn[rownames(dfn),]
ftn <- ftn[rownames(dfn),]
dfn$LibrarySize <- rowSums(stn)
ggplot(data=dfn, aes(x=BodySite, y=LibrarySize, color=BodySite)) + geom_jitter()
```

There are a number of low library samples, but we don't want to drop them all as they are useful for identifying contaminants. Not that many of the negative controls fall in this category! However a low minimum is still often appropriate:
```{r}
summary(rowSums(stn))
head(sort(rowSums(stn)),40)
# Requiring at least 30 reads
dfn <- dfn[rowSums(stn)>=30,]
stn <- stn[rownames(stn),]
ftn <- ftn[rownames(ftn),]
```

Add Picogreen DNA quantitation data:
```{r}
dfq <- read.csv(file.path(path, "DNA_quant_IL2_3_5-9_forBen_20170507.csv"), header=TRUE, stringsAsFactors = FALSE)
dfq <- dfq[match(rownames(dfn), dfq$specimen.barcode),]
identical(dfq$specimen.barcode, rownames(dfn)) # TRUE
dfn$Concentration <- dfq$ng.ul
```

Make phyloseq object:
```{r}
psn <- phyloseq(otu_table(ftn, taxa_are_rows=FALSE), sample_data(dfn), tax_table(taxn))
psn
```

Identify contaminants:
```{r}
cdf <- isContaminant(psn, conc="Concentration", neg="Negative", method="combined", threshold=0.1, batch="Run", batch.combine="fisher")
hist(cdf$p, 100)
```

Looks bimodal, with a peak at zero, just like we would hope.

Let's check raw numbers:
```{r}
table(cdf$contaminant)/nrow(cdf) # 2%
tapply(colSums(stn), cdf$contaminant, sum)/sum(stn) # 2% of reads
```

Just under 2% of the ASVs, and 2% of the reads, have been identified as contaminants.

Visualize 10 identified contaminants:
```{r}
set.seed(100)
i <- sample(which(cdf$contaminant), 10)
unname(cbind(paste0("Seq", i), taxn[i,]))
plot_frequency(psn, taxa_names(psn)[i], conc="Concentration", neg="Negative")
```

Frequency profiles and taxonomic assignments look consistent with contaminant origin.

## Revisit Exploratory Associations

This code was derived from the analysis workflows associated with Callahan & Digiluio et al. 2017.

Make a subject data.frame with the average frequencies for all genera from the processed data used in the manuscript:
```{r}
load(file.path(path, "processed.rda"))
subdf <- df[!duplicated(df$SubjectID),c("SubjectID", "Cohort", "Race", "preterm", "Outcome")]
subdf$Outcome <- factor(subdf$Outcome, levels=c("Preterm","Term"))
for(sq in rownames(tax)) {
  subdf[,sq] <- tapply(ft[,sq], df$SubjectID, mean)[subdf$SubjectID]
}
```

Test the association of the average gestational frequency of each amplicon sequence variant (ASV) with PTB:
```{r}
sqs <- rownames(tax)
scores.uab <- rep(1.0, nrow(tax))
scores.stan <- rep(1.0, nrow(tax))
names(scores.stan) <- sqs; names(scores.uab) <- sqs
for(sq in rownames(tax)) {
  scores.stan[sq] <- suppressWarnings(wilcox.test(subdf[subdf$Cohort %in% "Stanford" & subdf$preterm, sq],
                                 subdf[subdf$Cohort %in% "Stanford" & !subdf$preterm, sq], alternative="greater")$p.value)
  scores.uab[sq] <- suppressWarnings(wilcox.test(subdf[subdf$Cohort %in% "UAB" & subdf$preterm, sq],
                                subdf[subdf$Cohort %in% "UAB" & !subdf$preterm, sq], alternative="greater")$p.value)
}
taxdf <- data.frame(pStanford = scores.stan, pUAB = scores.uab, Genus=tax[, "Genus"])
taxdf$freqStanford <- apply(ft[df$Cohort %in% "Stanford",], 2, mean)
taxdf$freqUAB <- apply(ft[df$Cohort %in% "UAB",], 2, mean)
rownames(taxdf) <- sqs

taxdf$Joint <- pchisq(-2*log(taxdf$pStanford * taxdf$pUAB), df=4, lower.tail=FALSE)
taxdf$Sig <- p.adjust(taxdf$Joint, method="BH") < 0.1
taxdf$Freq <- (taxdf$freqStanford + taxdf$freqUAB)/2
sum(p.adjust(scores.stan, method="BH")<0.1); sum(p.adjust(scores.uab, method="BH")<0.1)
```

Propagate the decontam Ps to the exploratory analysis data.frame:
```{r}
rownames(cdf) <- colnames(stn)
taxdf$P.Contam <- cdf[rownames(taxdf), "p"]
```

Make exploratory plots (i.e. all variants) at the level of sequence variants, with points colored by their decontam P:
```{r}
taxdf$P.Contam[is.na(taxdf$P.Contam)] <- 0.5
taxdf$Category <- cut(taxdf$P.Contam, c(0, 1e-5, 0.01, 0.1, 1))
p.sv <- ggplot(data=taxdf, aes(x=pStanford, y=pUAB, color=Category, label=Genus)) + 
  geom_point() + 
  geom_text(data=taxdf[taxdf$P.Contam < 1e-6 & (taxdf$pUAB < 0.001 | taxdf$pStanford < 0.001),], color="magenta", vjust=-0.7) +
  scale_color_manual(values=c("(0,1e-05]"="magenta", "(1e-05,0.01]"="violetred2", "(0.01,0.1]"="grey", "(0.1,1]"="black"), name="decontam P") +
  scale_x_log10() + scale_y_log10() +
  theme_bw() + coord_fixed() + xlab("PTB Association P-value (Stanford)") + ylab("PTB Association P-value (UAB)")
p.sv
# ggsave(file.path(path.out, "PTB_contam.pdf"), p.sv, dev="pdf", width=6, height=4, units="in", useDingbats=FALSE)
```

Clearly several of the ASVs significantly associated with PTB in the exploratory analysis were contaminants, as suggested in the original manuscript.
