---
title: "Simple statistical identification and removal of contaminant sequences in marker-gene and metagenomics data: Oral analysis"
author: "Nicole Davis"
date: "11/02/2017"
output: html_document
---

This script contains the code required to reproduce Figures 1, 2, and S2. It generates the data featured in Table S1.

#prepare workspace
```{r, warning=FALSE, message=FALSE}
#load packages
cran_packages <- c("reshape2", "ggplot2", "vegan", "stringr", "gridExtra", "ape", "RColorBrewer", "plyr", "dplyr", "knitr","cowplot","scales")
bioc_packages <- c("phyloseq","decontam")
sapply(c(cran_packages, bioc_packages), library, character.only = TRUE)

#establish settings
theme_set(theme_bw())
options(stringsAsFactors = FALSE)
wd <- "~/Desktop/Contamination/Analyses/OralContamination" # CHANGE ME

#load phyloseq object
load(file.path(wd, 'MUC_ps.Rdata'))
MUC
MUCrelabund <- transform_sample_counts(MUC, function(x) x/sum(x))
```

#extract OTU / TAX tables, mapping file
```{r, echo=FALSE}
#extract tables
MAP <- data.frame(sample_data(MUC))
OTU <- data.frame(otu_table(MUC))
TAX <- data.frame(tax_table(MUC))

#add prevalence and ID columns to TAX table
TAX$id <- rownames(TAX)
TAX$prevalence <- colSums(OTU > 0)
TAX$abundance_abs <- taxa_sums(MUC)
```

#Figure 1
#contaminant and oral taxon abundances across samples
```{r}
#prepare Figure 1a, expected frequency patterns
x <- 1:100
y <- -1*x
sampletype <- c(rep('Biological', 50),rep('Control',50))

df <- data.frame(x=x, Contaminant=y, Non.Contaminant=rep(-1*50,100), sampletype=sampletype)
df2 <- melt(df, id.vars = c('x','sampletype'))
df2$variable <- as.character(df2$variable)
df2$variable[df2$variable == 'Non.Contaminant'] <- 'Non-Contaminant'
df2$slope <- -1
df2$slope[101:200] <- 0
df2$intercept <- -1
df2$intercept[101:200] <- -50

set.seed(2) #required for reproducible geom_jitter plot
plot1 <- ggplot(df2, aes(x,value))
f1a <- plot1 + 
  geom_jitter(height = 20, size=2) +
  geom_abline(aes(intercept=intercept, slope=slope, color=variable)) + 
  facet_wrap(~variable) +
  labs(x='log(DNA concentration)', y='log(Frequency)') +
  guides(color=FALSE) +
  scale_color_manual(values=c('red','blue')) +
  theme(panel.grid.minor = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),  
        axis.title.x = element_text(colour="grey20",size=10,angle=0,
                                    hjust=.5,vjust=0,face="plain"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),  
        axis.title.y = element_text(colour="grey20",size=10,angle=90,
                                    hjust=.5,vjust=.5,face="plain"),
        strip.text.x = element_text(colour = "grey20", size=8))

#Prepare Figure 1b, Prevalence patterns
set.seed(1618)
x1 <- sample(5:95, 100, replace=TRUE)/100
y1 <- sample(5:95, 100, replace=TRUE)/100
df <- data.frame(x=x1,y=y1)
df <- df[abs(df$x - df$y) > 0.1,]

df$SeqType <- 'Contaminant'
df$SeqType[df$y > df$x] <- 'Non-Contaminant'
df <- df[sample(nrow(df), 10), ]
df$SeqName <- paste0('Seq',1:dim(df)[1])
plot2 <- ggplot(df, aes(x,y, label=SeqName))

f1b <- plot2 + labs(x='Prevalence in Controls', y='Prevalence in Samples') + 
  scale_x_continuous(limits=c(0,1)) +
  scale_y_continuous(limits=c(0,1)) +
  geom_abline(slope=1, color='gray', linetype='dashed') + 
  geom_text(aes(color=SeqType)) + 
  theme(axis.text.x = element_text(colour="grey20",size=10,angle=0,                                hjust=0.55,vjust=.5,face="plain"),
        axis.text.y = element_text(colour="grey20",size=10,angle=0,
                                   hjust=1,vjust=0,face="plain"),  
        axis.title.x = element_text(colour="grey20",size=10,angle=0,
                                    hjust=.5,vjust=0,face="plain"),
        axis.title.y = element_text(colour="grey20",size=10,angle=90,
                                    hjust=.5,vjust=.5,face="plain"),
        legend.title = element_blank(),
        legend.text = element_text(size = 10, angle = 0),
        legend.position = 'bottom')

plot_grid(f1a,f1b, labels='auto',rel_widths=c(1.5,1))
```

#Figure 2a: Frequency patterns of taxa in the oral dataset
```{r}
#plot frequency of taxa
plot_abundance <- function(ps, taxa_to_plot, conc, taxa_are_rows=TRUE, norm=TRUE, log=TRUE, returndf=FALSE){

  ot <- as(otu_table(ps), "matrix")
    if(taxa_are_rows){
    ot <- t(ot)
  }
  
  taxa_mismatch <- !(taxa_names(ps) %in% colnames(ot))
  if(sum(taxa_mismatch) > 0){
    stop("Error: 'taxa_are_rows' argument may not be correct")
  }
  
  if(norm){
  ot <- sweep(ot, 1, rowSums(ot), "/")
  }
  
  ot <- as(ot[,colnames(ot) %in% taxa_to_plot], "matrix")
  if(dim(ot)[2] == 1){
    colnames(ot) <- taxa_to_plot
  }
    
  st <- as(sample_data(ps), "data.frame")
  snames <- sample_names(ps)
                   
  plot <- merge(st, ot, by.x = "row.names", by.y = "row.names", sort=FALSE)
  plot_melt <- melt(plot, id.vars = colnames(plot)[1:(dim(plot)[2]-length(taxa_to_plot))])

  colnames(plot_melt)[dim(plot_melt)[2]-1] <- "taxa_to_plot"
  colnames(plot_melt)[dim(plot_melt)[2]] <- "taxon_abundance"
  
  taxon_levels <- taxa_to_plot
  plot_melt$taxa_to_plot <- factor(plot_melt$taxa_to_plot, levels = taxon_levels)
  
  I <- which(colnames(plot_melt) == conc)
  colnames(plot_melt)[I] <- "DNA_conc"
  
  if(returndf == FALSE){
  if(log==TRUE){
    p1 <- ggplot(plot_melt, aes(log(DNA_conc),log(taxon_abundance)))
    return(p1 + geom_point())
  } else if(log==FALSE){
    p1 <- ggplot(plot_melt, aes(DNA_conc,taxon_abundance))
    return(p1 + geom_point())
  }
  } else if(returndf == TRUE){
    return(plot_melt)
  }
}


f2a <- plot_abundance(MUC,c('Seq3','Seq152','Seq53','Seq1','Seq12','Seq200'),"quant_reading",taxa_are_rows=FALSE) + facet_wrap(~taxa_to_plot,nrow=1) + labs(x= 'log(DNA concentration)', y = 'log(Seq Abundance)', color='', size=2) + geom_point(aes(color=Sample_or_Control)) +
   theme(axis.text.x = element_text(colour="grey20",size=15,angle=0,
                                    hjust=.5,vjust=.5,face="plain"),
        axis.text.y = element_text(colour="grey20",size=15,angle=0,
                                   hjust=1,vjust=0,face="plain"),  
        axis.title.x = element_text(colour="grey20",size=15,angle=0,
                                    hjust=.5,vjust=0,face="plain"),
        axis.title.y = element_text(colour="grey20",size=15,angle=90,
                                    hjust=.5,vjust=.5,face="plain"),
        strip.text.x = element_text(size = 15, angle = 0),
        legend.title = element_text(size = 15, angle = 0),
        legend.text = element_text(size = 15, angle = 0),
        legend.position = 'bottom')

f2a
```

#Figure 2b, Figure S2: examine p-values of taxa in oral dataset
```{r}
#Apply a minimal prevalence filter to filter out taxa present in fewer than 2 samples.
MUC_prev2samp <- filter_taxa(MUC, function(x) sum(x > 0) > 1, TRUE)
OTU2samp <- OTU[,colnames(OTU) %in% taxa_names(MUC_prev2samp)]

#load annotated taxonomy table. The column 'GenusREC' is the recommended taxonomy (Methods).
TAXannotate <- read.csv(file.path(wd, "TAXannotate.csv"))

for(i in 1:dim(TAXannotate)[1]){
  I = which(colnames(TAXannotate) == TAXannotate$Taxonomy_recommendation[i])
  if(length(I) ==0){
    TAXannotate$GenusREC[i] <- 'neither'
  } else
  TAXannotate$GenusREC[i] <- TAXannotate[i,I]
}

TAXsums <- data.frame(TAXsums = taxa_sums(MUC_prev2samp))
TAXprev <- data.frame(TAXprev = TAX$prevalence, Id = TAX$id)
TAXannotate <- merge(TAXannotate, TAXsums, by.x = 'Id', by.y = 'row.names', all.x=TRUE, all.y=FALSE, sort=FALSE)
TAXannotate <- merge(TAXannotate, TAXprev, by.x = 'Id', by.y = 'Id', all.x=TRUE, all.y=FALSE, sort=FALSE)

TAXa_formerging <- TAXannotate[c(1,2,9:18)]

#merge TAX with TAXannotate
TAX <- merge(TAX, TAXa_formerging, by.x = "id", by.y = "Id", all.x = TRUE, all.y = FALSE, sort=FALSE)

#Load database files
contam_database <- read.csv(file.path(wd, "contamination_database.csv"))
oral_database <- read.csv(file.path(wd, "oral_database.csv"))
HOMD <- read.csv(file.path(wd, "homd_taxonomy_table.csv"))

#Derive a list of genera that have been identified as contaminants in previous studies in the literature. This is the 'blacklist'.
contaminatinggenera <- contam_database %>% 
  group_by(Genus) %>% 
  summarize(numberstudies = length(unique(Reference)))

#Generate a list of genera that have been visualized in the mouth. Even if these taxa appear in the blacklist, we will not consider them contaminants.
orallist <- unique(oral_database$Genus)

#make blacklist. prevent overlap between orallist and blacklist by removing orallist taxa from blacklist
blacklist <- contaminatinggenera$Genus[contaminatinggenera$Genus != 'unspecified']
blacklist_nooral <- blacklist[!(blacklist %in% orallist)]

#Call contaminants using isContaminant, merge resulting p-values with TAXannotate
ocf <- isContaminant(as.matrix(OTU2samp), MAP$quant_reading, threshold = 0.1, detailed=TRUE, normalize=TRUE, method='frequency')

ocp <- isContaminant(as.matrix(OTU2samp), MAP$quant_reading, threshold = 0.1, detailed=TRUE, normalize=TRUE, method='prevalence',neg=MAP$Sample_or_Control == 'Control Sample')

occ <- isContaminant(as.matrix(OTU2samp), MAP$quant_reading, threshold = 0.1, detailed=TRUE, normalize=TRUE, method='combined',neg=MAP$Sample_or_Control == 'Control Sample')

pvalcols <- data.frame(id=rownames(ocf),pval.f=ocf$p, pval.p=ocp$p, pval.c = occ$p)

#merge those results with the TAXannotate dataframe
TAXannotate <- merge(TAXannotate, pvalcols, by.x = 'Id', by.y = 'id', all.x = TRUE, sort=FALSE)

#get p-values
get_p <- function(df, blacklist, orallist, thresholds, taxcolumn){
  
  df <- df[is.na(df$p) == FALSE,]
  
  all_cn <- vector('numeric', length(thresholds))
  all_cp <- vector('numeric', length(thresholds))
  all_cnReads <- vector('numeric', length(thresholds))
  all_cpReads <- vector('numeric', length(thresholds))
  
  ocn <- vector('numeric', length(thresholds))
  ocnReads <- vector('numeric', length(thresholds))
  bcn <- vector('numeric', length(thresholds))
  bcnReads <- vector('numeric', length(thresholds))
  
  ocp <- vector('numeric', length(thresholds))
  ocpReads <- vector('numeric', length(thresholds))
  bcp <- vector('numeric', length(thresholds))
  bcpReads <- vector('numeric', length(thresholds))
  
  #percent contaminating taxa and reads that are in blacklist, orallist
  tpo <- vector('numeric', length(thresholds))
  rpo <- vector('numeric', length(thresholds))
  tpb <- vector('numeric', length(thresholds))
  rpb <- vector('numeric', length(thresholds))
  
  gens <- vector('numeric', length(thresholds))
  
  G = which(colnames(df) == taxcolumn)

  for (i in 1:length(thresholds)){
    #what is number of isContaminants==TRUE overall?
    all_cn[i] <- sum(df$p <= thresholds[i])
    
    #percent isContaminants==TRUE overall?
    all_cp[i] <- sum(df$p <= thresholds[i]) / dim(df)[1]
    
    #num isContaminants==TRUE in lists?
    ocn[i] <- sum(df$p <= thresholds[i] & df[,G] %in% orallist)
    bcn[i] <- sum(df$p <= thresholds[i] & df[,G] %in% blacklist)
    
    #percent isContaminants=TRUE in lists?
    ocp[i] <- sum(df$p <= thresholds[i] & df[,G] %in% orallist) / dim(df)[1]
    bcp[i] <- sum(df$p <= thresholds[i] & df[,G] %in% blacklist) / dim(df)[1]
    
    ##repeat for read counts and percents
    
    #num isContaminants==TRUE overall?
    all_cnReads[i] <- sum(df$TAXsums[df$p <= thresholds[i]])
    
    #percent isContaminants==TRUE overall?
    all_cpReads[i] <- sum(df$TAXsums[df$p <= thresholds[i]]) / sum(df$TAXsums)
    
    #num isContaminants==TRUE in lists?
    ocnReads[i] <- sum(df$TAXsums[df$p <= thresholds[i] & df[,G] %in% orallist])
    bcnReads[i] <- sum(df$TAXsums[df$p <= thresholds[i] & df[,G] %in% blacklist])
    
    #percent isContaminants==TRUE in lists?
    ocpReads[i] <- sum(df$TAXsums[df$p <= thresholds[i] & df[,G] %in% orallist]) /
      sum(df$TAXsums)
    bcpReads[i] <- sum(df$TAXsums[df$p <= thresholds[i] & df[,G] %in% blacklist]) /
      sum(df$TAXsums)
    
    #number of unique genera at each threshold
    gens[i] <- length(unique(df[,G][df$p >= thresholds[i]]))
    
    #percent of contaminants found in whitelist
    tpo[i] <- ocn[i] / sum(df$p <= thresholds[i])
    tpb[i] <- bcn[i] / sum(df$p <= thresholds[i])
    
    #percent isContaminants=TRUE in lists?
    rpo[i] <- ocnReads[i] / sum(df$TAXsums[df$p <= thresholds[i]])
    rpb[i] <- bcnReads[i] / sum(df$TAXsums[df$p <= thresholds[i]])
    
  }
  
  #orallist
  df1 <- data.frame(Taxonomy = taxcolumn, 
                    Threshold=thresholds, 
                    NumberTaxa=ocn,
                    PercentTaxa=ocp,
                    PercentContamTaxainList = tpo,
                    PercentContamReadsinList = rpo,
                    NumberReads = ocnReads,
                    PercentReads = ocpReads,
                    List='Orallist',
                    Contam='isContaminant=TRUE')
  
  #blacklist
  df2 <- data.frame(Taxonomy = taxcolumn, 
                    Threshold=thresholds, 
                    NumberTaxa=bcn,
                    PercentTaxa=bcp,
                    PercentContamTaxainList = tpb,
                    PercentContamReadsinList = rpb,
                    NumberReads = bcnReads,
                    PercentReads = bcpReads,
                    List = 'Blacklist',
                    Contam='isContaminant=TRUE')
  
  #all
  df3 <- data.frame(Taxonomy = taxcolumn, 
                    Threshold=thresholds, 
                    NumberTaxa=all_cn,
                    PercentTaxa=all_cp,
                    PercentContamTaxainList = NA,
                    PercentContamReadsinList = NA,
                    NumberReads = all_cnReads,
                    PercentReads = all_cpReads,
                    List = 'All taxa',
                    Contam='isContaminant=TRUE')
  
  df4 <- rbind(df1,df2,df3)
  df4$UniqueGenera <- gens
  return(df4)
  
}
#test at multiple thresholds
thresholds <- seq(0, 1, length.out = 101)

##Plot p-value distributions
#set color palette
cbbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7","#000000")

#get p-values
TAXannotate$pval <- TAXannotate$pval.c

TAXannotate_full <- TAXannotate
TAXannotate_full$Group <- "All ASVs"

TAXannotate_oral <- TAXannotate
TAXannotate_oral$pval.c[!(TAXannotate$GenusREC %in% orallist)] <- NA
TAXannotate_oral$Group <- "Oral Database"

TAXannotate_contam <- TAXannotate
TAXannotate_contam$pval.c[!(TAXannotate$GenusREC %in% blacklist_nooral)] <- NA
TAXannotate_contam$Group <- "Contamination Database"

TAXann <- rbind(TAXannotate_full, TAXannotate_oral, TAXannotate_contam)

#2b
histo <- ggplot(TAXann, aes(pval.c, fill=Group))
f2b <- histo + geom_histogram() + labs(x = 'P-value', y='Number ASVs') + facet_wrap(~Group) + scale_fill_manual(values=cbbPalette) +
   theme(axis.text.x = element_text(colour="grey20",size=15,angle=0,
                                    hjust=.5,vjust=.5,face="plain"),
        axis.text.y = element_text(colour="grey20",size=15,angle=0,
                                   hjust=1,vjust=0,face="plain"),  
        axis.title.x = element_text(colour="grey20",size=15,angle=0,
                                    hjust=.5,vjust=0,face="plain"),
        axis.title.y = element_text(colour="grey20",size=15,angle=90,
                                    hjust=.5,vjust=.5,face="plain"),
        strip.text.x = element_text(size = 15, angle = 0),
        legend.title = element_text(size = 15, angle = 0),
        legend.text = element_text(size = 15, angle = 0),
        legend.position = 'bottom')

f2b

#combine 2a and 2b
plot_grid(f2a, f2b, labels='auto',nrow=2)

#S2, without log 
s2 <- ggplot(TAXann, aes(pval.c, fill=Group))
fs2a <- s2 + 
  geom_histogram(aes(weight=TAXsums)) + 
  labs(x = 'P-value', y='Number Reads') + 
  scale_y_continuous(labels=scientific) +
  facet_wrap(~Group) + 
  scale_fill_manual(values=cbbPalette) +
  guides(fill=FALSE) +
  theme(axis.text.x = element_text(colour="grey20",size=15,angle=0,
                                    hjust=.5,vjust=.5,face="plain"),
        axis.text.y = element_text(colour="grey20",size=15,angle=0,
                                   hjust=1,vjust=0,face="plain"),  
        axis.title.x = element_text(colour="grey20",size=15,angle=0,
                                    hjust=.5,vjust=0,face="plain"),
        axis.title.y = element_text(colour="grey20",size=15,angle=90,
                                    hjust=.5,vjust=.5,face="plain"),
        strip.text.x = element_text(size = 15, angle = 0))

fs2a

s2 <- ggplot(TAXann, aes(pval.c, fill=Group))
fs2b <- s2 + 
  geom_histogram(aes(weight=TAXsums)) + 
  labs(x = 'P-value', y='Number Reads') + 
  facet_wrap(~Group) + 
  scale_fill_manual(values=cbbPalette) +
  scale_y_log10() +
  
   theme(axis.text.x = element_text(colour="grey20",size=15,angle=0,
                                    hjust=.5,vjust=.5,face="plain"),
        axis.text.y = element_text(colour="grey20",size=15,angle=0,
                                   hjust=1,vjust=0,face="plain"),  
        axis.title.x = element_text(colour="grey20",size=15,angle=0,
                                    hjust=.5,vjust=0,face="plain"),
        axis.title.y = element_text(colour="grey20",size=15,angle=90,
                                    hjust=.5,vjust=.5,face="plain"),
        strip.text.x = element_text(size = 15, angle = 0),
        legend.title = element_text(size = 15, angle = 0),
        legend.text = element_text(size = 15, angle = 0),
        legend.position = 'bottom')

fs2b

plot_grid(fs2a,fs2b, labels='auto', nrow=2,rel_heights=c(1,1.25))
```

#Table S1: Blacklist taxa with high p-values
```{r}
#what are the blacklisters with high p-values?
TAXtmp <- TAXannotate[is.na(TAXannotate$pval.c) == FALSE,]
TAXtmp$GenusREC[TAXtmp$pval.c > 0.9 & TAXtmp$GenusREC %in% blacklist_nooral]
unique(TAXtmp$GenusREC[TAXtmp$pval.c > 0.9 & TAXtmp$GenusREC %in% blacklist_nooral])

oralmaybes <- TAXtmp$Id[TAXtmp$pval.c > 0.9 & TAXtmp$GenusREC %in% blacklist_nooral]

TAXannotate_oralmaybes <- TAXannotate[TAXannotate$Id %in% oralmaybes,]
#write.csv(TAXannotate_oralmaybes, 'oralmaybes.csv')
#this .csv file served as the basis for Table S1

plot_abundance(MUC, oralmaybes, "quant_reading", taxa_are_rows=FALSE) +
  geom_point(aes(color=Sample_or_Control)) +
  facet_wrap(~taxa_to_plot) + 
  labs(title='Blacklist taxa with high p-values')
```
